File: packages/react-native/android/src/main/java/dev/oneam/midnightmobile/MidnightMobileRuntimeModule.kt
package dev.oneam.midnightmobile

import android.util.Base64
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import dev.oneam.midnightmobile.uniffi.MidnightRuntimeException
import dev.oneam.midnightmobile.uniffi.applySyncBatch as runtimeApplySyncBatch
import dev.oneam.midnightmobile.uniffi.beginCommand as runtimeBeginCommand
import dev.oneam.midnightmobile.uniffi.cancelOperation as runtimeCancelOperation
import dev.oneam.midnightmobile.uniffi.closeWalletSession as runtimeCloseWalletSession
import dev.oneam.midnightmobile.uniffi.exportWalletCheckpoint as runtimeExportWalletCheckpoint
import dev.oneam.midnightmobile.uniffi.getWalletSnapshot as runtimeGetWalletSnapshot
import dev.oneam.midnightmobile.uniffi.openWalletSession as runtimeOpenWalletSession
import dev.oneam.midnightmobile.uniffi.resumeOperation as runtimeResumeOperation

private fun decodeBase64(value: String): ByteArray {
  val decoded = try {
    Base64.decode(value, Base64.NO_WRAP)
  } catch (error: IllegalArgumentException) {
    throw MidnightRuntimeException.InvalidArgument()
  }
  if (Base64.encodeToString(decoded, Base64.NO_WRAP) != value) {
    decoded.fill(0)
    throw MidnightRuntimeException.InvalidArgument()
  }
  return decoded
}

private fun decodeULong(value: Double): ULong {
  if (!value.isFinite() || value < 0 || value % 1.0 != 0.0 ||
    value > 9_007_199_254_740_991.0
  ) {
    throw MidnightRuntimeException.InvalidArgument()
  }
  return value.toLong().toULong()
}

private fun runtimeException(error: Throwable): CodedException {
  val code = when (error) {
    is MidnightRuntimeException.Unavailable -> "UNAVAILABLE"
    is MidnightRuntimeException.InvalidArgument -> "INVALID_ARGUMENT"
    is MidnightRuntimeException.StaleSession -> "STALE_SESSION"
    is MidnightRuntimeException.Cancelled -> "CANCELLED"
    is MidnightRuntimeException.StateIncompatible -> "STATE_INCOMPATIBLE"
    is MidnightRuntimeException.SyncGap -> "SYNC_GAP"
    is MidnightRuntimeException.ProofFailed -> "PROOF_FAILED"
    is MidnightRuntimeException.InsufficientDust -> "INSUFFICIENT_DUST"
    is MidnightRuntimeException.SubmissionStatusUnknown -> "SUBMISSION_STATUS_UNKNOWN"
    is MidnightRuntimeException.InvalidLength -> "INVALID_LENGTH"
    is MidnightRuntimeException.DecodeFailed -> "DECODE_FAILED"
    is MidnightRuntimeException.NativeInternal -> "NATIVE_INTERNAL"
    else -> "NATIVE_INTERNAL"
  }
  return CodedException(code, code, error)
}

class MidnightMobileRuntimeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MidnightMobileRuntime")

    AsyncFunction("openWalletSession") {
        configJson: String,
        nightExternalKey: ByteArray,
        zswapSeed: ByteArray,
        dustSeed: ByteArray,
        checkpointBase64: String? ->
      var checkpoint: ByteArray? = null
      try {
        checkpoint = checkpointBase64?.let(::decodeBase64)
        val value = runtimeOpenWalletSession(
          configJson,
          nightExternalKey,
          zswapSeed,
          dustSeed,
          checkpoint
        )
        mapOf("id" to value.id.toDouble(), "generation" to value.generation.toDouble())
      } catch (error: Throwable) {
        throw runtimeException(error)
      } finally {
        nightExternalKey.fill(0)
        zswapSeed.fill(0)
        dustSeed.fill(0)
        checkpoint?.fill(0)
      }
    }

    AsyncFunction("applySyncBatch") {
        sessionId: Double,
        generation: Double,
        stream: String,
        fromOffset: Double,
        toOffset: Double,
        payloadsBase64: List<String> ->
      var payloads = emptyList<ByteArray>()
      try {
        payloads = payloadsBase64.map(::decodeBase64)
        runtimeApplySyncBatch(
          decodeULong(sessionId),
          decodeULong(generation),
          stream,
          decodeULong(fromOffset),
          decodeULong(toOffset),
          payloads
        )
      } catch (error: Throwable) {
        throw runtimeException(error)
      } finally {
        payloads.forEach { it.fill(0) }
      }
    }

    AsyncFunction("getWalletSnapshot") {
        sessionId: Double, generation: Double ->
      try {
        runtimeGetWalletSnapshot(decodeULong(sessionId), decodeULong(generation))
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }

    AsyncFunction("exportWalletCheckpoint") {
        sessionId: Double, generation: Double ->
      try {
        val bytes = runtimeExportWalletCheckpoint(
          decodeULong(sessionId),
          decodeULong(generation)
        )
        try {
          Base64.encodeToString(bytes, Base64.NO_WRAP)
        } finally {
          bytes.fill(0)
        }
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }

    AsyncFunction("beginCommand") {
        sessionId: Double, generation: Double, commandJson: String ->
      try {
        runtimeBeginCommand(decodeULong(sessionId), decodeULong(generation), commandJson)
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }

    AsyncFunction("resumeOperation") {
        operationId: Double, generation: Double, networkResultJson: String? ->
      try {
        runtimeResumeOperation(
          decodeULong(operationId),
          decodeULong(generation),
          networkResultJson
        )
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }

    AsyncFunction("cancelOperation") {
        operationId: Double, generation: Double ->
      try {
        runtimeCancelOperation(decodeULong(operationId), decodeULong(generation))
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }

    AsyncFunction("closeWalletSession") {
        sessionId: Double, generation: Double ->
      try {
        runtimeCloseWalletSession(decodeULong(sessionId), decodeULong(generation))
      } catch (error: Throwable) {
        throw runtimeException(error)
      }
    }
  }
}

File: packages/react-native/ios/MidnightMobileRuntimeModule.swift
import ExpoModulesCore
import Foundation

private func runtimeException(_ error: Error) -> Exception {
  let code: String
  switch error {
  case MidnightRuntimeError.Unavailable:
    code = "UNAVAILABLE"
  case MidnightRuntimeError.InvalidArgument:
    code = "INVALID_ARGUMENT"
  case MidnightRuntimeError.StaleSession:
    code = "STALE_SESSION"
  case MidnightRuntimeError.Cancelled:
    code = "CANCELLED"
  case MidnightRuntimeError.StateIncompatible:
    code = "STATE_INCOMPATIBLE"
  case MidnightRuntimeError.SyncGap:
    code = "SYNC_GAP"
  case MidnightRuntimeError.ProofFailed:
    code = "PROOF_FAILED"
  case MidnightRuntimeError.InsufficientDust:
    code = "INSUFFICIENT_DUST"
  case MidnightRuntimeError.SubmissionStatusUnknown:
    code = "SUBMISSION_STATUS_UNKNOWN"
  case MidnightRuntimeError.InvalidLength:
    code = "INVALID_LENGTH"
  case MidnightRuntimeError.DecodeFailed:
    code = "DECODE_FAILED"
  case MidnightRuntimeError.NativeInternal:
    code = "NATIVE_INTERNAL"
  default:
    code = "NATIVE_INTERNAL"
  }
  return Exception(name: "MidnightRuntimeError", description: code, code: code)
}

private func decodeBase64(_ value: String) throws -> Data {
  guard let data = Data(base64Encoded: value),
        data.base64EncodedString() == value else {
    throw MidnightRuntimeError.InvalidArgument
  }
  return data
}

private func decodeUInt64(_ value: Double) throws -> UInt64 {
  guard value.isFinite, value >= 0, value.rounded(.towardZero) == value,
        value <= 9_007_199_254_740_991 else {
    throw MidnightRuntimeError.InvalidArgument
  }
  return UInt64(value)
}

public class MidnightMobileRuntimeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MidnightMobileRuntime")

    AsyncFunction("openWalletSession") {
      (
        configJson: String,
        nightExternalKeyBytes: Data,
        zswapSeedBytes: Data,
        dustSeedBytes: Data,
        checkpointBase64: String?
      ) -> [String: Double] in
      var nightExternalKey = nightExternalKeyBytes
      var zswapSeed = zswapSeedBytes
      var dustSeed = dustSeedBytes
      var checkpoint = Data()
      do {
        defer {
          nightExternalKey.resetBytes(in: 0..<nightExternalKey.count)
          zswapSeed.resetBytes(in: 0..<zswapSeed.count)
          dustSeed.resetBytes(in: 0..<dustSeed.count)
          checkpoint.resetBytes(in: 0..<checkpoint.count)
        }
        checkpoint = try checkpointBase64.map(decodeBase64) ?? Data()
        let value = try openWalletSession(
          configJson: configJson,
          nightExternalKey: nightExternalKey,
          zswapSeed: zswapSeed,
          dustSeed: dustSeed,
          checkpoint: checkpointBase64 == nil ? nil : checkpoint
        )
        return ["id": Double(value.id), "generation": Double(value.generation)]
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("applySyncBatch") {
      (
        sessionId: Double,
        generation: Double,
        stream: String,
        fromOffset: Double,
        toOffset: Double,
        payloadsBase64: [String]
      ) -> String in
      var payloads = [Data]()
      do {
        defer {
          for index in payloads.indices {
            payloads[index].resetBytes(in: 0..<payloads[index].count)
          }
        }
        payloads = try payloadsBase64.map(decodeBase64)
        return try applySyncBatch(
          sessionId: decodeUInt64(sessionId),
          generation: decodeUInt64(generation),
          stream: stream,
          fromOffset: decodeUInt64(fromOffset),
          toOffset: decodeUInt64(toOffset),
          payloads: payloads
        )
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("getWalletSnapshot") {
      (sessionId: Double, generation: Double) -> String in
      do {
        return try getWalletSnapshot(
          sessionId: decodeUInt64(sessionId),
          generation: decodeUInt64(generation)
        )
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("exportWalletCheckpoint") {
      (sessionId: Double, generation: Double) -> String in
      do {
        let checkpoint = try exportWalletCheckpoint(
          sessionId: decodeUInt64(sessionId),
          generation: decodeUInt64(generation)
        )
        return checkpoint.base64EncodedString()
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("beginCommand") {
      (sessionId: Double, generation: Double, commandJson: String) -> String in
      do {
        return try beginCommand(
          sessionId: decodeUInt64(sessionId),
          generation: decodeUInt64(generation),
          commandJson: commandJson
        )
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("resumeOperation") {
      (
        operationId: Double,
        generation: Double,
        networkResultJson: String?
      ) -> String in
      do {
        return try resumeOperation(
          operationId: decodeUInt64(operationId),
          generation: decodeUInt64(generation),
          networkResultJson: networkResultJson
        )
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("cancelOperation") {
      (operationId: Double, generation: Double) in
      do {
        try cancelOperation(
          operationId: decodeUInt64(operationId),
          generation: decodeUInt64(generation)
        )
      } catch {
        throw runtimeException(error)
      }
    }

    AsyncFunction("closeWalletSession") {
      (sessionId: Double, generation: Double) in
      do {
        try closeWalletSession(
          sessionId: decodeUInt64(sessionId),
          generation: decodeUInt64(generation)
        )
      } catch {
        throw runtimeException(error)
      }
    }
  }
}
